### PR TITLE
LGR support

### PR DESCRIPTION
**Pull request by sunlelma**
LGR support with image transformation, error checking, open/close, etc.

### CODE REVIEW

**Comment by sunlelma** [comment]
I feel like the .py files should be organized by level/replay/lgr as practically speaking that is how they would be used (instead of modules, packing, etc)

Anywho. Tada. LGRs.

**Comment by sunlelma** [comment]
Markku also points out that I mispelt transparency (as transpErAncy)

**Comment by sigvef** [comment]
LGR support, awesome! Do you have documentation for the LGR format somewhere? I just have a txt file with level and replay formats.

> I feel like the .py files should be organized by level/replay/lgr as practically speaking that is how they would be used (instead of modules, packing, etc)

Yeah, that makes sense. I can restructure after this pull request gets merged.

**Comment by sunlelma** [comment]
I made this:

http://wiki.elmaonline.net/LGR/Technical

I have a few more corrections/additions to this pull request after

File: elma/error.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            else:
+                message.append([ERR_INVALID_PALETTE,lgro,"%s has an invalid palette"%(lgro.name)])
+            if(namelower=="qcolors"):
+                if(lgro.img.width!=LGR_WARNING_QCOLORS_WIDTH or lgro.img.height!=LGR_WARNING_QCOLORS_HEIGHT):
+                    message.append([WARN_QCOLOR_DIMENSIONS,lgro,"Warning: qcolors usually has dimensions of %sx%s, but the dimensions are %sx%s"%(LGR_WARNING_QCOLORS_WIDTH,LGR_WARNING_QCOLORS_HEIGHT,lgro.img.width,lgro.img.height)])
+            if(namelower=="q1bike" or namelower=="q2bike"):
+                if(lgro.img.width<LGR_RECOMMEND_BIKE_WIDTH or lgro.img.height<LGR_RECOMMEND_BIKE_HEIGHT):
+                    message.append([WARN_BIKE_SMALL,lgro,"Warning: %s needs dimensions of at least %sx%s to be rendered properly, but the dimensions are %sx%s"%(lgro.name,LGR_RECOMMEND_BIKE_WIDTH,LGR_RECOMMEND_BIKE_HEIGHT,lgro.img.width,lgro.img.height)])
+        else:
+            message.append([ERR_IMG_MISSING,lgro,"%s has no image!"%(lgro.name)])
```
Comment (sigvef) [inline elma/error.py]: `ERR_IMG_MISSING` is not defined. Is it a typo perhaps?

**Comment by sunlelma** [comment]
Yeah I guess I didn't test the error function as well as I should have :).

The following errors were tested/produced:
ERR_LGR_INVALID_PALETTE
ERR_MISSING_MANDATORY_PCX
ERR_DUPLICATE_NAME
WARN_UNUSED_FOOD
ERR_IMAGE_TYPE
ERR_DISTANCE
ERR_TRANSPARENCY
ERR_TOO_WIDE
ERR_TOO_LARGE
ERR_OBJ_WIDTH
ERR_INVALID_PALETTE
WARN_OBJ_HEIGHT
WARN_PALETTE_MISMATCH

These errors I'm not sure if I fully tested
ERR_TOO_MANY_GRASS
ERR_TOO_MANY_TEXTURES
ERR_NOT_ENOUGH_TEXTURES
ERR_TOO_MANY_PICTURES
ERR_TOO_MANY_MASKS
ERR_OBJ_TOO_WIDE
ERR_SMALL_IMAGE_ONLY
ERR_IMG_MISSING
WARN_GRASS_HEIGHT
WARN_QCOLOR_DIMENSIONS
WARN_BIKE_SMALL

Should maybe test these errors in lgr_test py

**Comment by sunlelma** [comment]
Let me add some stuff to lgr_test.py

**Comment by sunlelma** [comment]
Done! Very comprehensive error-checking of my functions. Found a few errors thanks to that :)

Take a look again please :)

**Comment by sigvef** [comment]
Very nice, will review it again!

**Comment by sigvef** [comment]
Looks good, thank you! 🥇

**Pull request merged**